protocols/kad/src/behaviour.rs

```diff
@@ -71,188 +65,110 @@ pub struct Kademlia<TSubstream, TRecordStorage: RecordStore = MemoryRecordStorag
-        match &self.inner {
-            QueryInfoInner::Initialization { target } => KademliaHandlerIn::FindNodeReq {
-                key: target.clone().into(),
-                user_data,
-            },
-            QueryInfoInner::FindPeer(key) => KademliaHandlerIn::FindNodeReq {
-                // TODO: Change the `key` of `QueryInfoInner::FindPeer` to be a `Multihash`.
-                key: key.clone().into(),
-                user_data,
-            },
-            QueryInfoInner::GetProviders { target, .. } => KademliaHandlerIn::GetProvidersReq {
-                key: target.clone(),
-                user_data,
-            },
-            QueryInfoInner::AddProvider { .. } => KademliaHandlerIn::FindNodeReq {
-                key: unimplemented!(), // TODO: target.clone(),
-                user_data,
-            },
-            QueryInfoInner::GetValue { key, .. } => KademliaHandlerIn::GetValue {
-                key: key.clone(),
-                user_data,
-            },
-            QueryInfoInner::PutValue { key, .. } => KademliaHandlerIn::FindNodeReq {
-                key: key.clone(),
-                user_data,
-            }
-        }
+    /// Sets a custom protocol name.
+    ///
+    /// Kademlia nodes only communicate with other nodes using the same protocol name. Using a
+    /// custom name therefore allows to segregate the DHT from others, if that is desired.
+    pub fn set_protocol_name(&mut self, name: impl Into<Cow<'static, [u8]>>) -> &mut Self {
+        self.protocol_name_override = Some(name.into());
+        self
     }
-}
 
-impl<TSubstream, TRecordStorage> Kademlia<TSubstream, TRecordStorage>
-where
-    TRecordStorage: RecordStore
-{
-    /// Creates a new `Kademlia` network behaviour with the given local `PeerId`.
-    pub fn new(local_peer_id: PeerId) -> Self
-    where
-        TRecordStorage: Default
-    {
-        Self::new_inner(local_peer_id, Default::default())
+    /// Sets the record storage implementation to use.
+    pub fn set_storage(&mut self, store: TStore) -> &mut Self {
+        self.records = Some(store);
+        self
     }
 
-    /// The same as `new`, but using a custom protocol name.
+    /// Sets the timeout for a single query.
     ///
-    /// Kademlia nodes only communicate with other nodes using the same protocol name. Using a
-    /// custom name therefore allows to segregate the DHT from others, if that is desired.
-    pub fn with_protocol_name(local_peer_id: PeerId, name: impl Into<Cow<'static, [u8]>>) -> Self
-    where
-        TRecordStorage: Default
-    {
-        let mut me = Kademlia::new_inner(local_peer_id, Default::default());
-        me.protocol_name_override = Some(name.into());
-        me
+    /// > **Note**: A single query usually comprises at least as many requests
+    /// > as the replication factor, i.e. this is not a request timeout.
+    ///
+    /// The default is 60 seconds.
+    pub fn set_query_timeout(&mut self, timeout: Duration) -> &mut Self {
```

Comment:
Why have a query timeout at all? The query will end at some point, and an extra timeout just means that we might fail if our network is poor and the value is far away.

Reply:
Running queries take up resources and all resource use should be bounded in a controlled fashion. While every query should "end at some point" if everything works correctly, that "at some point" can be a very, _very_ long time. If a user is willing to let queries run that long he can still do so by configuring a very large timeout. Furthermore, that a query "ends at some point" without a timeout depends on quite a few implementation details, some of which out of the control of `libp2p-kad`, namely that all the callbacks for connected / disconnected / dialing failed / etc. are all always invoked correctly for all peers participating in a query without a single mistake, otherwise there is already a memory leak that may go unnoticed for a long time. Explicit, hard timeouts put a bound here and make such issues more visible through the produced timeout errors.

In this manner, explicit timeouts provide some control over how many queries may actually be still running, e.g. if you want to issue a random query every minute say, as in substrate, having a query timeout of 1 minute ensures that only one such query is running at a time (or at most 2 for a timeout of 2 minutes, at most 3 for a timeout of 3 minutes or whatever is desired).